wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/CommandScheduler.java

```diff
@@ -141,8 +141,8 @@ public void clearButtons() {
    */
   private void initCommand(Command command, boolean interruptible, Set<Subsystem> requirements) {
     CommandState scheduledCommand = new CommandState(interruptible);
-    m_scheduledCommands.put(command, scheduledCommand);
     command.initialize();
+    m_scheduledCommands.put(command, scheduledCommand);
     for (Subsystem requirement : requirements) {
       m_requirements.put(requirement, command);
     }
```

Comment:
I believe the problem here is that this code change means that `cancel` has no effect within `initialize`, but the test cases assert that it will successfully cancel the command.  Putting the initialize at the end should mean that `cancel` has full effect.

My only hesitation here is that the original ordering of the code may have been motivated by the idea that the requirements of a command are not frozen on construction, but can be modified within `initialize`.  I have no idea if that was intended, or if it would even work.  I'm also unsure why anyone would want to do that.

```suggestion
    m_scheduledCommands.put(command, scheduledCommand);
    for (Subsystem requirement : requirements) {
      m_requirements.put(requirement, command);
    }
   command.initialize();
```

Reply:
> My only hesitation here is that the original ordering of the code may have been motivated by the idea that the requirements of a command are not frozen on construction, but can be modified within initialize. I have no idea if that was intended, or if it would even work. I'm also unsure why anyone would want to do that.

Command requirements aren't something that should be modified dynamically.

---
> I believe the problem here is that this code change means that cancel has no effect within initialize, but the test cases assert that it will successfully cancel the command. Putting the initialize at the end should mean that cancel has full effect.

Good catch! Now that I think about it again, it seems that this might be a good way to solve [the schedule-from-end recursion issue](https://github.com/wpilibsuite/allwpilib/pull/2450) from a while ago. Do we want to reorder calling `end()` to *after* removing the command from the scheduled commands collection?

Reply:
> Command requirements aren't something that should be modified dynamically.

Agreed, but we don't document that, and we don't enforce it.

Reply:
> Moving the call to end to the end would probably fix that, although it seems a little icky that end can call commands on a subsystem the command doesn't currently own.

That's my hesitation as well.

> An alternative solution here would be to defer both schedule and cancel within cancel and initCommand as if we were in the run loop.

Not sure I understand what you mean?

Reply:
> Not sure I understand what you mean?

Inside `CommandScheduler.run`, a key section has the `m_inRunLoop` variable set to `true`.  The effect of this is that if either `schedule` or `cancel` is called within `execute` or `isFinished`, they won't act immediately, but will instead push the commands onto the `m_toSchedule` and `m_toCancel` queues.  They will then be scheduled and cancelled appropriately after the section ends.  This was put in in [PR 1938](https://github.com/wpilibsuite/allwpilib/pull/1938) to "fixes concurrentmodificationexception when commands are scheduled/canceled from the execute body of other commands".  

The problem we're facing here is very similar.  If `schedule` or `cancel` is called within `initialize` or `end`, we want it to take effect, but without corrupting data-structures.  We can probably get away with doing this by careful ordering of data-structure updates, but I'm suggesting that a similar pattern of deferring `schedule` and `cancel` to a safe, consistent point might be more clearly correct and robust.  Maybe a couple of methods like `startDeferringScheduleAndCancel` and `stopDeferringScheduleAndCancel` that generalize PR 1938 would help.

Reply:
It's a very minor break.  It bugs me that it's invisible, but also any team code that actually breaks based on this was probably *very brittle*.

Reply:
I haven't verified this, but don't all the unit tests rely on initialize being called from schedule?

Reply:
> So you mean having _all_ commands be added to the `m_toSchedule`/`m_toCancel` collection, and then the actual scheduling/cancel be during the run loop?

Yes (or mostly), to the first part, but I would clear those collections at the end of `initCommand` and `cancel`, not wait for the end of the next `run` command.  That was my intent behind the suggestion of having two methods:

```
void startDeferringScheduleAndCancel() {
  m_inRunLoop = true; // consider renaming this boolean
}

void stopDeferringScheduleAndCancel() {
  m_inRunLoop = false;

  // Schedule/cancel commands from queues populated during loop
  for (Map.Entry<Command, Boolean> commandInterruptible : m_toSchedule.entrySet()) {
    schedule(commandInterruptible.getValue(), commandInterruptible.getKey());
  }

  for (Command command : m_toCancel) {
    cancel(command);
  }

  m_toSchedule.clear();
  m_toCancel.clear();
}
```

These could be used in pairs in all three of `run`, `initCommand` and `cancel`, and would make it much clearer what is going on.

Reply:
> any team code that actually breaks based on this was probably _very brittle_

Agreed.  Any time I've seen a student trying to use `cancel` and `schedule`, it turns out that there is a better way to organize the code.

Reply:
I actually think it's a good idea to do the actual initializing/canceling at a specified point in the run loop; I think it reduces the chances of race conditions. Might cause some other issues though, that's why I'm a bit apprehensive about it.

Reply:
There's a latent bug in the current implementation of deferred `schedule`/`cancel` that would be carried over into my proposal: The deferred operations may be re-ordered.  If a user calls `cancel` then `schedule` within `execute` or `isFinished` (current code) or within `initialize` or `end` (my proposal), then it will be executed as `schedule`, then `cancel`, resulting in the command being unscheduled, whereas the user would have expected it to be scheduled.  

This could be mitigated by merging the deferred `schedule` and `cancel` lists into a single command queue, but I'm not sure it's worth it.  I would never recommend that a student write client code like this.  The important thing is that the scheduler maintains consistent data-structures in these cases and doesn't crash.